### PR TITLE
C++: Clean up BufferWrite.qll

### PR DESCRIPTION
**Pull request by geoffw0**
While working on the QLDoc for BufferWrite.qll, I noticed some places where it could depend more on models (reducing redundancy), and where models can be improved using cases from BufferWrite.qll.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/implementations/Strcat.qll [hunk 1]
```diff
@@ -24,6 +24,21 @@ class StrcatFunction extends TaintFunction, DataFlowFunction, ArrayFunction, Sid
     )
   }
 
+  /**
+   * Gets the index of the parameter that is the size of the copy (in characters).
+   */
+  int getParamSize() { if exists(getParameter(2)) then result = 2 else none() }
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Strcat.qll:30]: Writing this as `exists(getParameter(2)) and result = 2` feels more QL'y to me.

**Comment by MathiasVP** [comment]
Other than the failing ql{doc, format} checks, I think this LGTM! The new test result also looks right to me.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Strcat.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Strcat.qll]: Agreed.  Updated.

**Comment by geoffw0** [comment]
Fixed formatting and fixed the new false positive in `UsingStrcpyAsBoolean.ql`.  It was caused by the query assuming that all `strcpy` variants return a string, whereas some of the new ones (e.g. `strcpy_s`) do in fact return an error code so the query doesn't apply to them.

**Pull request merged**